### ISSUE TITLE
applying `mqtt::buffer` to `message` and APIs such as `publish`

### ISSUE THREAD

**Issue by redboltz** [redboltz/mqtt_cpp#338]
I'd like to consider applying `mqtt::buffer` widely after #320 is merged.

Now, `message` contains `as::buffer`. It is a kind of view type. So we need to store lifetime object separately. But if `message` contains `mqtt::buffer` instead of `as::buffer`, it can contains the lifetime object.

In addition, APIs interface could be changed to `mqtt::buffer`. I guess  that the `mqtt::buffer` move into `message`.

I thihk that it is a good improvement.

NOTE: `mqtt::buffer` can share the one big buffer as follows:

```
   +-----------------------------------------+
   |                                         |
   V                                         |
shared_count <--- buffer1                  buffer2
   |                |                        |
   |                |                        |
   V                V                        V
   +-------+--------+------------+-----------+---------+
   |header | length | topic_name | packet_id | payload | 
   +-------+--------+------------+-----------+---------+
```

**Comment by jonesmz** [comment]
This is a good idea.

It's important that we not remove the existing boost::asio::buffer functions, because in the situation where the data is from one large buffer, only one life-keeper is needed.

If we remove the versions of the functions that take boost::asio::buffer and a life-keeper, then we force the end-user to accept double the amount of reference counting.

**Comment by redboltz** [comment]
> It's important that we not remove the existing boost::asio::buffer functions, because in the situation where the data is from one large buffer, only one life-keeper is needed.

Yes, they should be kept.

**Comment by redboltz** [comment]
> If we remove the versions of the functions that take boost::asio::buffer and a life-keeper, then we force the end-user to accept double the amount of reference counting.

Let me clarify the amount of reference counting.

Focused on async_publish with QoS1.

* as::buffer + life_keeper
  * user creates life_keeper object. It is typically `std::pair<std::shared_ptr<std::string>, std::shared_ptr<std::string>>`.
  * In `async_send_publish()` function, 
    * life_keeper is copied into `store_`. At this point, ref-count is incremented.
    * life_keeper is move captured to `do_async_write()` handler.
* mqtt::buffer without life_keeper
  * mqtt::buffer contains shared_ptrs.
  * In `async_send_publish()` function, 
    * mqtt::buffer is moved into `basic_publish_message` (variable name `msg`).
    * `msg` is copied to `store_msg`. At this point, ref-count is incremented.
    * `store_msg` is moved into `store_`.
    * `msg` is moved to `do_async_write()`.

I think that they are the same amount of reference counter operations.

If user do as follows, then the amount of reference counting is doubled.

* mqtt::buffer with life_keeper
  * mqtt::buffer contains shared_ptrs in addition life_keeper also contains shared_ptrs.

Do you think so?

**Comment by redboltz** [comment]
Implemented by #364
